<?xml version="1.0"?>

<robot name="fwdis" xmlns:xacro="http://ros.org/wiki/xacro">

  <xacro:include filename="$(find four_wheel_drive_independent_steering)/urdf/base/base.urdf.xacro"/>
  <xacro:include filename="$(find four_wheel_drive_independent_steering)/urdf/wheel/wheel.urdf.xacro"/>
  <xacro:include filename="$(find four_wheel_drive_independent_steering)/urdf/steering/steering.urdf.xacro"/>
  <xacro:include filename="$(find four_wheel_drive_independent_steering)/urdf/sensors/lrf.urdf.xacro"/>

  <!-- Link & Joint -->
  <link name="base_footprint"/>

  <xacro:base parent="base_footprint">
    <origin xyz="0 0 0.2"/>
  </xacro:base>

  <xacro:steering prefix="front_right" parent="base_link">
    <origin xyz="0.25 -0.25 0.12" rpy="0 0 0"/>
    <axis xyz="0 0 1"/>
  </xacro:steering>

  <xacro:wheel prefix="front_right" parent="front_right_steering_link">
    <origin xyz="0 0 -0.15" rpy="${M_PI/2} 0 0"/>
    <axis xyz="0 0 -1"/>
  </xacro:wheel>

  <xacro:steering prefix="front_left" parent="base_link">
    <origin xyz="0.25 0.25 0.12" rpy="0 0 0"/>
    <axis xyz="0 0 1"/>
  </xacro:steering>

  <xacro:wheel prefix="front_left" parent="front_left_steering_link">
    <origin xyz="0 0 -0.15" rpy="${M_PI/2} 0 0"/>
    <axis xyz="0 0 -1"/>
  </xacro:wheel>

  <xacro:steering prefix="rear_right" parent="base_link">
    <origin xyz="-0.25 -0.25 0.12" rpy="0 0 0"/>
    <axis xyz="0 0 1"/>
  </xacro:steering>

  <xacro:wheel prefix="rear_right" parent="rear_right_steering_link">
    <origin xyz="-0 0 -0.15" rpy="${M_PI/2} 0 0"/>
    <axis xyz="0 0 -1"/>
  </xacro:wheel>

  <xacro:steering prefix="rear_left" parent="base_link">
    <origin xyz="-0.25 0.25 0.12" rpy="0 0 0"/>
    <axis xyz="0 0 1"/>
  </xacro:steering>

  <xacro:wheel prefix="rear_left" parent="rear_left_steering_link">
    <origin xyz="-0 0 -0.15" rpy="${M_PI/2} 0 0"/>
    <axis xyz="0 0 -1"/>
  </xacro:wheel>

  <xacro:hokuyo prefix="lrf" parent="base_link">
    <origin xyz="0 0 1" rpy="0 0 0"/>
  </xacro:hokuyo>

  <!-- Transmission -->
  <xacro:wheel_trans prefix="front_right"/>
  <xacro:wheel_trans prefix="front_left"/>
  <xacro:wheel_trans prefix="rear_right"/>
  <xacro:wheel_trans prefix="rear_left"/>
  <xacro:steering_trans prefix="front_right"/>
  <xacro:steering_trans prefix="front_left"/>
  <xacro:steering_trans prefix="rear_right"/>
  <xacro:steering_trans prefix="rear_left"/>

  <!-- Gazebo -->
  <gazebo>
    <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
      <robotNamespace>/fwdis</robotNamespace>
    </plugin>
  </gazebo>
  <xacro:base_gazebo/>
  <xacro:wheel_gazebo prefix="front_right"/>
  <xacro:wheel_gazebo prefix="front_left"/>
  <xacro:wheel_gazebo prefix="rear_right"/>
  <xacro:wheel_gazebo prefix="rear_left"/>
  <xacro:steering_gazebo prefix="front_right"/>
  <xacro:steering_gazebo prefix="front_left"/>
  <xacro:steering_gazebo prefix="rear_right"/>
  <xacro:steering_gazebo prefix="rear_left"/>
  <xacro:hokuyo_gazebo prefix="lrf" min_rad="-1.65806278939" max_rad="1.65806278939" min_range="0.10" max_range="60.0"/>
</robot>
